# 가설검정

## 이론

다음과 같은 두 가지 가설을 생각한다.

- 귀무가설 (null hypothesis) $\text{H}_0$ : 모집단에 대하여 기존에 사실이라고 알려져 있거나 처음부터 주장하는 가설

- 대립가설 (alternative hypothesis) $\text{H}_1$ : 귀무가설에 대립되어 만들어지는 가설로 모집단에 대한 주장을 담고 있으며 실험이나 조사를 통해 입증하고 싶은 가설

가설검정은 조사된 데이터의 통계량(표본평균, 표본분산 등)을 통하여 귀무가설을 기각하고 대립가설을 채택할 충분한 증거가 있는지 확인하는 과정으로 다음과 같은 두 가지의 오류를 생각할 수 있다.

  제 1종 오류: 귀무가설이 참인데 귀무가설을 기각할 오류. 제1종 오류를 범할 확률은 보통 $\alpha$로 나타내고 이를 유의수준 (Significance Level)이라 한다.  

  제 2종 오류: 귀무가설이 거짓인데 귀무가설을 채택할 오류. 제2종 오류를 범할 확률은 보통 $\beta$로 나타내는데 제 1종 오류를 범할 확률 $\alpha$를 높이면 $\beta$는 낮아지는 경향이 있다.

- 제 1종 오류를 허용할 한계인 유의수준 $\alpha$ 만을 다루기로 한다. 유의수준은 연구자가 설정하는 값으로, 일반적으로 0.05 (5%) 또는 0.01(1%)를 많이 사용한다.  유의수준이 낮을수록 제1종 오류를 범할 확률이 줄어든다.

**기각역과 가설검정의 종류**

유의수준($\alpha$)과 대립가설의 성격에 따라 귀무가설을 기각하는 영역(기각역)이 정해지며 이에 따라 다음과 같은 종류의 가설검정을 생각할 수 있다.
*   **양측검정 (Two-tailed test)**: 대립가설이 '차이가 있다' ($\textbf{H}_1: \mu \ne \mu_0$) 와 같이 특정 값과 다르다는 형태일 때 사용한다.  이때는 분포의 양쪽 꼬리 부분에 기각역이 설정되며, 유의수준 $\alpha$는 양쪽 꼬리에 $\alpha/2$씩 나누어진다.
*   **단측검정 (One-tailed test)**: 대립가설이 '더 크다' ($\textbf{H}_1: \mu > \mu_0$) 또는 '더 작다' ($\textbf{H}_1: \mu < \mu_0$) 와 같이 특정 방향의 차이를 주장할 때 사용한다. 이때는 분포의 한쪽 꼬리에만 기각역이 설정되며, 유의수준 $\alpha$ 전체가 해당 꼬리 부분에 위치한다.

**유의수준과 임계값(critical value)**

표준정규분포 $N(0, 1^2)$을 따르는 확률변수 $Z$에 대하여

$$P(Z\ge z_{\alpha})=\alpha $$

라 하면 확률 $\alpha$에 따른 변수 $z_{\alpha}$가 정해진다.

이때 주어진 유의수준 $\alpha$에 대한 가설검정의 기각역은 양측검정의 경우

$$\alpha = P(|Z| \ge z_{\alpha/2}) = P(Z\le -z_{\alpha/2}, Z \ge z_{\alpha/2})$$

단측검정의 경우

$$\alpha = P(Z\le -z_{\alpha}) \,\,또는 \,\,\alpha = P(Z\ge z_{\alpha})$$

이다.

- 여기서 유의수준 $\alpha$의 중요한 기준이 되는 임계값을 생각할 수 있는데  양측검정의 경우 두 개의 임계값 $-z_{\alpha/2}, z_{\alpha/2}$, 단측검정의 경우 한 개의 임계값 $-z_{\alpha}$ 또는 $z_{\alpha}$이 있다.
                 







**표본평균 분포**

확률 변수 $X$로 나타내어지는 모집단의 평균을 $\mu$ 분산을 $\sigma^2$이라 하자. 이때, 충분히 큰 크기 $n$인 표본의 표본평균 $\overline{X}$는 중심극한정리에 의하여 정규분포 $N(\mu, \sigma^2/n)$을 따른다. 즉,

$$\overline{X} \sim N(\mu, \sigma^2/n)$$

이라 표현한다. 또한, 정규분포 성질에 의하여

$$Z = \frac{\overline{X} - \mu}{\sigma/\sqrt{n}}\sim N(0, 1^2)$$

이다.
- 따라서 충분히 큰 표본으로부터의 모평균 관련한 추정과 검정은 표준정규분포를 사용한다.





**검정통계량과 $p$-값**

충분히 큰 크기 $n$의 표본 데이터를 가지고 모평균에 대한 가설검정을 하려 하며 표준정규분포를 사용한다.

귀무가설 $\text{H}_0$ : 모평균은 $\mu_0$이다.

- 검정통계량(Test Statistic):  표본 데이터로부터 계산되는 값으로 가설검정을 수행하는 기준이 된다. 이 경우 조사된 표본의 평균값을 $X_0$라 하면  

$$z_0= \frac{X_0 - \mu_0}{\sigma/\sqrt{n}}$$

이 검정통계량이 된다. (여기서 모분산이 알려지지 않은 경우 표본의 불편분산을 사용하기도 한다.)

- $p$-값($p$-value): 귀무가설 하에서 현재 검정통계량 값 또는 그보다 더 극단적인 값이 나타날 확률로

$$ P(z\le z_0) 또는  P(z\ge z_0)$$

인데 이 경우 표준정규분포를 이용하여 계산하고 대립가설 $\text{H}_1$에 따라 선택한다.

**귀무가설 기각**

유의수준 $\alpha$/임계값 그리고 조사된 데이터로부터의 $p$-값/검정추정량을 비교하여 가설검정을 진행한다.

방법 1. 임계값과 검정추정량 비교

- 검정통계량이 기각역에 속하면 귀무가설을 기각한다.
- 검정통계량이 기각역에 속하지 않으면 귀무가설을 기각할 수 없다.

방법 2. 유의수준과 $p$-값 비교

- 단측검정에서는 $p$-값이 유의수준 $\alpha$보다 작은 경우, 양측검정에서는 $p$-값이 $\alpha/2$보다 작은 경우 귀무가설을 기각한다.
- 그렇지 않으면 귀무가설을 기각할 수 없다.  

**가설검정 절차**

- 귀무가설 $\text{H}_0$, 대립가설 $\text{H}_1$ 설정
- 분포 결정
- 유의수준/임계값
- $p$-값/검정통계량 (표본 데이터로 부터 계산한다)
- 귀무가설을 기각할 지를 결정한다.

**예제 (단측 검정, 왼쪽꼬리)**

어떤 회사는 자사 제품의 평균 사용 시간이 100시간이라고 주장합니다.  새로운 공정으로 생산된 제품의 평균 사용 시간이 100시간보다 짧은지 유의수준 5%에서 검정해 봅시다.

30개의 표본을 추출하였더니 평균은 98이 왔음.

모분산을 $10^2$이라 가정합니다.

- **귀무가설 $\text{H}_0$**: 새로운 공정으로 생산된 제품의 평균 사용 시간은 100시간이다 ($\mu = 100$).
- **대립가설 $\text{H}_1$**: 새로운 공정으로 생산된 제품의 평균 사용 시간은 100시간보다 짧다 ($\mu < 100$).

이것은 단측 검정(왼쪽 꼬리)입니다.


**가설 검정 절차:**

1.  **귀무가설 $\text{H}_0$, 대립가설 $\text{H}_1$ 설정**:
    *   $\text{H}_0: \mu = 100$
    *   $\text{H}_1: \mu < 100$

2.  **분포 결정**: 모분산이 알려져 있고 표본 크기가 충분히 크므로 표준정규분포 $N(0, 1^2)$을 사용합니다.

3.  **유의수준/임계값**:
    *   유의수준 $\alpha = 0.05$ 입니다.
    *   단측 검정(왼쪽 꼬리)이므로, 표준정규분포에서 왼쪽 꼬리 확률이 0.05가 되는 z-값을 찾습니다.
    *   표준정규분포표 또는 계산기를 찾아보면, 누적 확률이 0.05가 되는 z-값은 약 -1.645 입니다.
    *   따라서 임계값은 $-z_{0.05} \approx -1.645$ 입니다.
    *   기각역은 $Z \le -1.645$ 입니다.

4.  **$p$-값/검정통계량 (표본 데이터로 부터 계산한다)**:
    *   주어진 정보: 표본평균 $X_0 = 98$, 모평균 (귀무가설 하) $\mu_0 = 100$, 모표준편차 $\sigma = 10$, 표본 크기 $n = 30$.
    *   검정통계량 $z_0$를 계산합니다.
        $$z_0= \frac{X_0 - \mu_0}{\sigma/\sqrt{n}} = \frac{98 - 100}{10/\sqrt{30}}$$
    *   $\sqrt{30} \approx 5.477$ 이므로
        $$z_0 \approx \frac{-2}{10/5.477} = \frac{-2}{1.826} \approx -1.095$$
    *   검정통계량은 약 -1.095 입니다.

    *   p-값을 계산합니다. 대립가설이 $\mu < 100$이므로, 귀무가설 하에서 검정통계량이 -1.095보다 작거나 같을 확률을 계산합니다.
        $$p\text{-값} = P(Z \le z_0) = P(Z \le -1.095)$$
    *   표준정규분포표 또는 계산기를 사용하면, $P(Z \le -1.095)$는 약 0.1367 입니다.

5.  **귀무가설을 기각할 지를 결정한다**:

    *   **방법 1. 임계값과 검정통계량 비교**:
        *   검정통계량 $z_0 \approx -1.095$ 입니다.
        *   임계값은 약 -1.645 입니다.
        *   검정통계량 $(-1.095)$은 임계값 $(-1.645)$보다 크므로 기각역($Z \le -1.645$)에 속하지 않습니다.
        *   따라서 귀무가설을 기각할 수 없습니다.

    *   **방법 2. 유의수준과 $p$-값 비교**:
        *   유의수준 $\alpha = 0.05$ 입니다.
        *   p-값 $\approx 0.1367$ 입니다.
        *   p-값 $(0.1367)$은 유의수준 $(0.05)$보다 크므로 귀무가설을 기각할 수 없습니다.

**결론**: 유의수준 5% 하에서 새로운 공정으로 생산된 제품의 평균 사용 시간이 100시간보다 짧다고 할 충분한 증거가 없습니다.

## 예제

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
pd.options.mode.copy_on_write = True

In [ ]:
%precision 6

'%.6f'

In [ ]:
pd.set_option('display.precision', 6)

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
data_url = 'https://raw.githubusercontent.com/codingalzi/DataSci/refs/heads/master/data/'

In [ ]:
housing = pd.read_csv(data_url+"california_housing.csv")
housing

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

In [ ]:
housing = housing.loc[:, ['median_income', 'median_house_value']]
housing

median_income  median_house_value
0             8.3252            452600.0
1             8.3014            358500.0
2             7.2574            352100.0
3             5.6431            341300.0
4             3.8462            342200.0
...              ...                 ...
20635         1.5603             78100.0
20636         2.5568             77100.0
20637         1.7000             92300.0
20638         1.8672             84700.0
20639         2.3886             89400.0

[20640 rows x 2 columns]

In [ ]:
house_value_max = housing['median_house_value'].max() # 500,001

mask = housing['median_house_value'] >= house_value_max

In [ ]:
housing = housing[~mask]
housing.shape

(19675, 2)

In [ ]:
housing = housing.reset_index(drop=True)
housing.index.name = 'district'
housing

median_income  median_house_value
district                                   
0                8.3252            452600.0
1                8.3014            358500.0
2                7.2574            352100.0
3                5.6431            341300.0
4                3.8462            342200.0
...                 ...                 ...
19670            1.5603             78100.0
19671            2.5568             77100.0
19672            1.7000             92300.0
19673            1.8672             84700.0
19674            2.3886             89400.0

[19675 rows x 2 columns]

In [ ]:
housing["income_cat"] = pd.cut(housing["median_income"],
                               bins=[0., 1.5, 3.0, 4.5, 6., np.inf],
                               labels=[1, 2, 3, 4, 5])

housing

median_income  median_house_value income_cat
district                                              
0                8.3252            452600.0          5
1                8.3014            358500.0          5
2                7.2574            352100.0          5
3                5.6431            341300.0          4
4                3.8462            342200.0          3
...                 ...                 ...        ...
19670            1.5603             78100.0          2
19671            2.5568             77100.0          2
19672            1.7000             92300.0          2
19673            1.8672             84700.0          2
19674            2.3886             89400.0          2

[19675 rows x 3 columns]

위 자료 housing에서 집값의 평균은 192477.921017이다.

In [ ]:
house_value_mean = housing['median_house_value'].mean()
house_value_mean

192477.921017

 각 소득분위별 집값에 대한 이야기를 하려 한다.

예를 들어 2소득분위에 속한 표본 66 개를 추출하여 2소득 분위에 속한 사람들이 산 집값이 평균보다 낮은지를 유의수준 5%(1%)로 판단하여라.

In [ ]:
mask = housing['income_cat'] == 2
sample_2Cat = housing[mask].sample(frac=0.01, random_state=42)
sample_2Cat

median_income  median_house_value income_cat
district                                              
2931             2.5644             73000.0          2
2128             2.3967             58700.0          2
11520            1.9531            162500.0          2
2144             2.7857             79700.0          2
17954            1.7411             98400.0          2
...                 ...                 ...        ...
13462            1.9500            150000.0          2
2773             2.7583             64900.0          2
4819             1.8086             97100.0          2
12269            2.3958             78400.0          2
14407            2.9000            185200.0          2

[66 rows x 3 columns]

표본평균은 129143.939394이다.

In [ ]:
sample_2Cat_house_value_mean = sample_2Cat.median_house_value.mean()
sample_2Cat_house_value_mean

129143.939394

귀무가설 $\text{H}_0$: `sample_2Cat_house_value_mean` = `house_value_mean`

대립가설 $\text{H}_1$: `sample_2Cat_house_value_mean` < `house_value_mean`

귀무가설이 참이라고 가정.

$$\overline{X} \sim N(\mu, \sigma^2/n)$$

$$Z = \frac{\overline{X} - \mu}{\sigma/\sqrt{n}}\sim N(0, 1^2)$$

*** 왼쪽꼬리검정을 시행한다.

In [ ]:
house_value_std = housing.median_house_value.std(ddof=0)
house_value_std

97709.026443

검정통계량 계산

In [ ]:
z_0 = (sample_2Cat_house_value_mean - house_value_mean) / (house_value_std / np.sqrt(len(sample_2Cat)))
z_0

-5.265918

반면에 유의수준 5% 임계값은 -1.644854이다.

In [ ]:
from scipy.stats import norm

`ppf()` 함수는 누적분포함수 `cdf()`의 역함수임.

In [ ]:
z_5p = norm.ppf(0.05)
z_5p

-1.644854

따라서 소득 2구간에서 임의로 선택된 66개 구역의 평균값은 전체 평균값보다 낮을 확률이 5% 미만이기에
귀무가설을 기각한다.

*** 왼쪽 꼬리 검정, 기각역은 $\{z
\le -1.644854 (\text{임계값)}\}$

*** 검정통계량 $z_0 = -5.265918 < -1.644854$ 이므로 기각역에 속한다.

*** 따라서 귀무가설을 기각한다.

결론: 소득 2구간에서만 표본을 선택하였기에 당연히 전체 평균값보다 작음.

*** 즉, 소득2분위에 속한 사람들이 산 집값이 평균보다 낮다고 볼 수 있다.  

**소득 3구간을 대상으로 하는 경우**

In [ ]:
mask = housing['income_cat'] == 3
sample_3Cat = housing[mask].sample(frac=0.01, random_state=43)

sample_3Cat_house_value_mean = sample_3Cat.median_house_value.mean()
sample_3Cat_house_value_mean

197321.126761

평균값보다 크다.

In [ ]:
house_value_mean

192477.921017

귀무가설 $\text{H}_0$: `sample_3Cat_house_value_mean` = `house_value_mean`

대립가설 $\text{H}_1$: `sample_3Cat_house_value_mean` > `house_value_mean`

귀무가설이 참이라고 가정.

$$\overline{X} \sim N(\mu, \sigma^2/n)$$

$$Z = \frac{\overline{X} - \mu}{\sigma/\sqrt{n}}\sim N(0, 1^2)$$

*** 오른쪽 꼬리 검정을 시행한다.

검정통계량이 0.417664이다.

In [ ]:
z_0 = (sample_3Cat_house_value_mean - house_value_mean) / (house_value_std / np.sqrt(len(sample_3Cat)))
z_0

0.417664

반면에 유의수준 5% 임계값은 1.644854이다.

In [ ]:
z_5p = norm.ppf(0.95)
z_5p

1.644854

*** 오른쪽 꼬리 검정, 기각역은 $\{z
\ge 1.644854 (\text{임계값)}\}$

*** 검정통계량 $z_0 = 0.417664 < 1.644854$ 이므로 기각역에 속하지 않는다.  

*** 따라서 귀무가설을 기각하지 않는다.


결론: 검정통계량이 임계값보다 작기에 귀무가설을 유지한다.

*** 즉, 소득3분위에 속한 사람들이 산 집값이 평균보다 높다고 볼 수 없다.